### PR TITLE
Import custom attributes with GLTFDocumentExtension

### PR DESCRIPTION
**Pull request by huwpascoe**
## Proposal
Import ARRAY_CUSTOM# attributes with GLTFDocumentExtension using a function to specify layer name. Also added "Export Attributes" checkbox to the blender importer.
https://github.com/godotengine/godot-proposals/issues/10892

## Test Project
image
[custom_attrib_test.zip](https://github.com/user-attachments/files/23870955/custom_attrib_test.zip)

## changes
- Ability to combine 2 attributes per CUSTOM
- Created new GLTFAttributeMap resource to deal with the sheer number of possibilities
```gdscript
extends GLTFDocumentExtension

func _import_get_attribute_map(state: GLTFState, mesh_index: int) -> GLTFAttributeMap:
	var attrmap := GLTFAttributeMap.new()

	var mesh_name: String = state.json["meshes"][mesh_index]["name"]
	match mesh_name:
		"mesh_1": attrmap.color = "COLOR_1"
		"mesh_2": attrmap.color = "_MY_DATA"
		"custom":
			attrmap.custom0 = "_ROCK"
			attrmap.custom1 = "_MUD"
			attrmap.custom2 = ""
			attrmap.custom3 = "_GRASS"
			attrmap.custom3_mux = "_SNOW"

	return attrmap
```

## Supported attributes
- Mesh.ARRAY_TEX_UV (vec2)
- Mesh.ARRAY_TEX_UV2 (vec2)
- Mesh.ARRAY_COLOR (vec3, vec4)
- Mesh.ARRAY_CUSTOM# (scalar, vec2, vec3, vec4)

### CODE REVIEW

**Comment by fire** [comment]
I'm having a hard time understanding the extension. Can you link or write a proposal in https://github.com/godotengine/godot-proposals?

Something like: I have a game workflow that uses blender I want the vertex attributes to import from blender, but it needs this pr. I can use the attributes for a shader effect.

File: modules/gltf/gltf_document.cpp [hunk 1]
```diff
@@ -3136,6 +3136,64 @@ Error GLTFDocument::_parse_meshes(Ref<GLTFState> p_state) {
...
+						case GLTFAccessor::TYPE_VEC2: {
+							array[custom_i] = _decode_accessor_as_vec2(p_state, a[attrib_name], true, indices_mapping);
+							format = Mesh::ARRAY_CUSTOM_RG_FLOAT;
+						} break;
+						case GLTFAccessor::TYPE_VEC3: {
+							array[custom_i] = _decode_accessor_as_vec3(p_state, a[attrib_name], true, indices_mapping);
+							format = Mesh::ARRAY_CUSTOM_RGB_FLOAT;
+						} break;
+						case GLTFAccessor::TYPE_VEC4: {
+							array[custom_i] = _decode_accessor_as_floats(p_state, a[attrib_name], true, indices_vec4_mapping);
```
Comment (aaronfranke) [inline modules/gltf/gltf_document.cpp:3170]: Why are Vec3s decoded as Vector3, but Vec4s are not decoded as Vector4?

File: modules/gltf/extensions/gltf_document_extension.cpp [hunk 2]
```diff
@@ -34,6 +34,7 @@ void GLTFDocumentExtension::_bind_methods() {
 	// Import process.
 	GDVIRTUAL_BIND(_import_preflight, "state", "extensions");
 	GDVIRTUAL_BIND(_get_supported_extensions);
+	GDVIRTUAL_BIND(_remap_custom_attribute, "name");
```
Comment (aaronfranke) [inline modules/gltf/extensions/gltf_document_extension.cpp:37]: This isn't documented, you need to run doctool.

File: modules/gltf/gltf_document.cpp (hunk 1 above)
Comment (huwpascoe) [reply modules/gltf/gltf_document.cpp:3170]: There's not a decode_as_vec4 method, that's how it's done in the rest of the class. But come to think of it, they probably all need to be decoded as float, thanks for pointing that out.

**Comment by huwpascoe** [comment]
https://github.com/godotengine/godot-proposals/issues/10892

**Comment by huwpascoe** [comment]
Refactored and removed like a hundred duplicate lines of code.
```cpp
_decode_accessor_as_floats();
_decode_accessor_as_ints();
_decode_accessor_as_vec2();
_decode_accessor_as_vec3();
_decode_accessor_as_color();
_decode_accessor_as_quaternion();
_decode_accessor_as_xform2d();
_decode_accessor_as_basis();
_decode_accessor_as_xform();

// VVV

_decode_accessor<T>();
```

Multi-element scalars are now decoded with _decode_accessor<T, N>()
```cpp
array[Mesh::ARRAY_TANGENT] = _decode_accessor<float, 4>(p_state, a["TANGENT"], true, indices_mapping);
```

**Comment by aaronfranke** [comment]
Since this PR alters the accessor code, it will have to wait until after https://github.com/godotengine/godot/pull/94165 which also touches this code.

**Comment by Flynsarmy** [comment]
> Since this PR alters the accessor code, it will have to wait until after #94165 which also touches this code.

PR in question was just merged so we should be good to continue here.

**Comment by fire** [comment]
@huwpascoe Many people think this is a good enhancement. Would you happen to have time to rebase it? <3

**Comment by huwpascoe** [comment]
@fire

Would it be better to keep as an extension or promote it to import options?
image

There are trade-offs for each.

**Comment by huwpascoe** [comment]
> Remember to take this PR out of draft if it's ready :)

Don't worry! I haven't forgotten.

> I am trending towards promoting it to import options

Right, I've determined that the UI approach is just too much of a mess for what is a technical feature. So I'll be moving forward with the original plan to add the method to DocumentExtension.

**Comment by huwpascoe** [comment]
I'll make a test project soon.

File: modules/gltf/gltf_document.h [hunk 3]
```diff
@@ -143,38 +143,15 @@ class GLTFDocument : public Resource {
 	Vector<double> _decode_accessor(Ref<GLTFState> p_state,
 			const GLTFAccessorIndex p_accessor,
 			const bool p_for_vertex);
-	Vector<float> _decode_accessor_as_floats(Ref<GLTFState> p_state,
-			const GLTFAccessorIndex p_accessor,
-			const bool p_for_vertex,
-			const Vector<int> &p_packed_vertex_ids = Vector<int>());
-	Vector<int> _decode_accessor_as_ints(Ref<GLTFState> p_state,
```
Comment (aaronfranke) [inline modules/gltf/gltf_document.h]: Apologies for blocking this PR for a second time, but this accessor code is already refactored by PR #109103, so it will have to wait until after that PR is merged.

File: modules/gltf/extensions/gltf_document_extension.h [hunk 4]
```diff
@@ -44,6 +44,7 @@ class GLTFDocumentExtension : public Resource {
 	// Import process.
 	virtual Error import_preflight(Ref<GLTFState> p_state, const Vector<String> &p_extensions);
 	virtual Vector<String> get_supported_extensions();
+	virtual void get_attribute_for_mesh_array(Ref<GLTFState> p_state, GLTFMeshIndex p_index, Mesh::ArrayType p_mesh_array, String &r_attribute);
```
Comment (aaronfranke) [inline modules/gltf/extensions/gltf_document_extension.h]: I know the existing API is a mix, but I would prefer prefixing all of the new import functions with `import_`.

More importantly, we should use the return value instead of `r_`.

```suggestion
	virtual String import_get_attribute_for_mesh_array(Ref<GLTFState> p_state, GLTFMeshIndex p_index, Mesh::ArrayType p_mesh_array);
```

Then in the loop below, it would store the result in local variables, check if the return type is not `.is_empty()`, and if so, set to the persistent variables.

Oh, also, the functions in `GLTFDocumentExtension` need to be listed in the order in which they are executed. So `import_get_attribute_for_mesh_array` needs to go between `parse_texture_json` and `import_object_model_property`.

File: modules/gltf/gltf_document.cpp [hunk 5]
```diff
@@ -1571,68 +1594,109 @@ Error GLTFDocument::_parse_meshes(Ref<GLTFState> p_state) {
...
-				if (!num_channels) {
-					break;
-				}
-				if (num_channels == 2 || num_channels == 4) {
-					cur_custom.resize(vertex_num * num_channels);
-					for (int32_t uv_i = 0; uv_i < texcoord_first.size() && uv_i < vertex_num; uv_i++) {
-						cur_custom.write[uv_i * num_channels + 0] = texcoord_first[uv_i].x;
-						cur_custom.write[uv_i * num_channels + 1] = texcoord_first[uv_i].y;
+			if (using_custom_attributes) {
+				for (int custom_i = 0; custom_i < 4; ++custom_i) {
```
Comment (aaronfranke) [inline modules/gltf/gltf_document.cpp]: Godot's style is to use postfix `++`.

```suggestion
				for (int custom_i = 0; custom_i < 4; custom_i++) {
```

File: modules/gltf/gltf_document.cpp [hunk 6]
```diff
@@ -1509,6 +1509,29 @@ Error GLTFDocument::_parse_meshes(Ref<GLTFState> p_state) {
...
+				ext->get_attribute_for_mesh_array(p_state, i, Mesh::ARRAY_COLOR, attr_color);
+				ext->get_attribute_for_mesh_array(p_state, i, Mesh::ARRAY_TEX_UV, attr_uv1);
+				ext->get_attribute_for_mesh_array(p_state, i, Mesh::ARRAY_TEX_UV2, attr_uv2);
+				ext->get_attribute_for_mesh_array(p_state, i, Mesh::ARRAY_CUSTOM0, attr_custom[0]);
+				ext->get_attribute_for_mesh_array(p_state, i, Mesh::ARRAY_CUSTOM1, attr_custom[1]);
+				ext->get_attribute_for_mesh_array(p_state, i, Mesh::ARRAY_CUSTOM2, attr_custom[2]);
+				ext->get_attribute_for_mesh_array(p_state, i, Mesh::ARRAY_CUSTOM3, attr_custom[3]);
+			}
+
+			for (String s : attr_custom) {
```
Comment (aaronfranke) [inline modules/gltf/gltf_document.cpp]:
```suggestion
			for (const String &s : attr_custom) {
```

File: modules/gltf/gltf_document.cpp [hunk 7]
```diff
@@ -1571,68 +1594,109 @@ Error GLTFDocument::_parse_meshes(Ref<GLTFState> p_state) {
...
+				for (int custom_i = 0; custom_i < 4; ++custom_i) {
+					const String &attr = attr_custom[custom_i];
+					if (attr.is_empty() || !a.has(attr)) {
+						continue;
+					}
+
+					Mesh::ArrayCustomFormat format{};
+					bool format_okay = true;
+
+					const Ref<GLTFAccessor> accessor = p_state->accessors[a[attr]];
```
Comment (aaronfranke) [inline modules/gltf/gltf_document.cpp]:
```suggestion
					const Ref<GLTFAccessor> &accessor = p_state->accessors[a[attr]];
```

File: modules/gltf/gltf_document.cpp [hunk 8]
```diff
@@ -1571,68 +1594,109 @@ Error GLTFDocument::_parse_meshes(Ref<GLTFState> p_state) {
...
-					// Vector.resize seems to not zero-initialize. Ensure all unused elements are 0:
-					for (int32_t uv_i = texcoord_first.size(); uv_i < vertex_num; uv_i++) {
-						cur_custom.write[uv_i * num_channels + 0] = 0;
-						cur_custom.write[uv_i * num_channels + 1] = 0;
+
+					if (!format_okay) {
+						Array arr;
+						arr.push_back(attr);
+						ERR_PRINT(String("Unknown format for {}, ignoring attribute.").format(arr));
+						continue;
```
Comment (aaronfranke) [inline modules/gltf/gltf_document.cpp]: Since only one place can set this to false, why not get rid of the bool and put this inside of `default: {}`?

**Comment by huwpascoe** [comment]
I've updated the test scene for the name change (`_import_get_attribute_for_mesh_array`)

File: modules/gltf/gltf_document.cpp [hunk 9]
```diff
@@ -1571,68 +1616,104 @@ Error GLTFDocument::_parse_meshes(Ref<GLTFState> p_state) {
...
-				}
-				if (!num_channels) {
-					break;
-				}
-				if (num_channels == 2 || num_channels == 4) {
-					cur_custom.resize(vertex_num * num_channels);
-					for (int32_t uv_i = 0; uv_i < texcoord_first.size() && uv_i < vertex_num; uv_i++) {
-						cur_custom.write[uv_i * num_channels + 0] = texcoord_first[uv_i].x;
-						cur_custom.write[uv_i * num_channels + 1] = texcoord_first[uv_i].y;
+			if (using_custom_attributes) {
```
Comment (lyuma) [inline modules/gltf/gltf_document.cpp]: I would like to see the code paths merged. That is, the builtin mapping of TEXCOORD attributes to custom should work the same way as the same mapping done as a custom mapping. (basically, the current Godot behavior should be implemented as a default "custom mapping" that could then be overridden.)

This means even if I only want to override one specific custom attribute, it means that I can no longer use the builtin mapping of TEXCOORD_2..9 into CUSTOM0..3 for the other channels, while UV and UV2 will still work the same

**Review by lyuma — COMMENTED** [review]
I have some comments about the design that I'd like a chance to discuss.

Firstly, I have concerns about the usability of depending on a customized GLTFDocumentExtension to perform the remapping... it certainly could work, but document extensions are registered globally (project-wide), so any attempt to use this feature would have to condition the remapping on some condition or add a setting to the import dock. If I were to copy the sample code as written, it would break all other glTF models in my project, even if they do not have that custom attribute. But explaining all this is probably too much for a method description, so I'm not sure how to clarify this.

Second, as I noticed, if I use this to override any custom attributes, then I can no longer combine two UV channels into one custom channel which limits the number of custom 2-channel UVs I can import. I feel like maybe something simple like "TEXCOORD_1,TEXCOORD_2" could work to handle that case. I think it only needs to support `2 x vec2` or `1 x anything` cases , so no need to handle every combination.

Basically the usecase is for compacting multiple 2-channel UVs into a CUSTOM, and this is Godot's current behavior so it would be nice to support it for that reason without needing a whole special case. At a minimum, it should be possible to return "" for an individual custom channel to get the current behavior without losing all the other custom channels.

**Comment by huwpascoe** [comment]
> If I were to copy the sample code as written, it would break all other glTF models in my project, even if they do not have that custom attribute. But explaining all this is probably too much for a method description, so I'm not sure how to clarify this.

That's why GLTFState is provided to determine if changes apply. If this method needs explaining, then every other method of GLTFDocumentExtension would require the same explanation. So that's out of scope, I think.

> At a minimum, it should be possible to return "" for an individual custom channel to get the current behavior without losing all the other custom channels.

That sounds reasonable, I'll see if I can get it working efficiently.

**Comment by huwpascoe** [comment]
Individual attribute querying is gone, now extensions only need to return a GLTFAttributeMap resource. With defaults set to how texture UVs were originally handled.

Any 2 attributes can combine into a CUSTOM provided they fit within a vec4. For example, vec3+scalar is valid.

image

File: modules/gltf/doc_classes/GLTFAttributeMap.xml [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+<class name="GLTFAttributeMap" inherits="Resource" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="../../../doc/class.xsd">
+	<brief_description>
+		Controls how GLTF attributes are mapped.
+	</brief_description>
+	<description>
+		Controls how GLTF attributes are mapped. to the resulting [Mesh]. Used by [method GLTFDocumentExtension._import_get_attribute_map]
+		[b]Note:[/b] It may be necessary to disable [code]generate_lods[/code] in the mesh import settings to preserve custom data.
+
+		Multiplexing is supported with CUSTOM* attributes. Setting custom* and custom*_mux will combine them into the one array.
+		The attributes must total no larger than vec4. If the _mux attribute is missing, it will act like only the first was set.
```
Comment (Calinou) [inline modules/gltf/doc_classes/GLTFAttributeMap.xml]:
```suggestion
		Controls how glTF attributes are mapped to the resulting [Mesh]. Used by [method GLTFDocumentExtension._import_get_attribute_map].
		Multiplexing is supported with [code]CUSTOM*[/code] attributes. Setting [code]custom*[/code] and [code]custom*_mux[/code] will combine them into the one array.
		A pair of [code]custom*[/code] and [code]custom*_mux[/code] attributes must total no larger than [code]vec4[/code]. For instance, you can set [member custom0] to a scalar and [member custom0_mux] to a [code]vec3[/code], but you cannot set [member custom0] to a [code]vec2[/code] and [member custom0_mux] to a [code]vec3[/code] (as this would require 5 values).
		If the [code]custom*_mux[/code] attribute that is requested is missing, it will act like only the first was set.
		[b]Note:[/b] It may be necessary to disable [code]generate_lods[/code] in the mesh import settings to preserve custom data.
```

File: modules/gltf/doc_classes/GLTFAttributeMap.xml [hunk 11]
```diff
@@ -0,0 +1,62 @@
+<?xml version="1.0" encoding="UTF-8" ?>
+<class name="GLTFAttributeMap" inherits="Resource" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="../../../doc/class.xsd">
+	<brief_description>
+		Controls how GLTF attributes are mapped.
```
Comment (Calinou) [inline modules/gltf/doc_classes/GLTFAttributeMap.xml]:
```suggestion
		Controls how glTF attributes are mapped on import.
```

File: modules/gltf/doc_classes/GLTFAttributeMap.xml [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+	<tutorials>
+	</tutorials>
+	<members>
+		<member name="color" type="String" setter="set_color" getter="get_color" default="&quot;COLOR_0&quot;">
+			Targets [constant Mesh.ARRAY_COLOR]
+			Supported attributes: vec3, vec4
+		</member>
+		<member name="custom0" type="String" setter="set_custom0" getter="get_custom0" default="&quot;TEXCOORD_2&quot;">
+			Targets [constant Mesh.ARRAY_CUSTOM0]
+			Supported attributes: scalar, vec2, vec3, vec4
```
Comment (Calinou) [inline modules/gltf/doc_classes/GLTFAttributeMap.xml]: It may be better to list scalar types explicitly. As I understand it, this would be `bool`, `int` or `float`?

File: modules/gltf/doc_classes/GLTFAttributeMap.xml [hunk 13]
```diff
@@ -0,0 +1,62 @@
...
+
+		Multiplexing is supported with CUSTOM* attributes. Setting custom* and custom*_mux will combine them into the one array.
+		The attributes must total no larger than vec4. If the _mux attribute is missing, it will act like only the first was set.
+	</description>
+	<tutorials>
+	</tutorials>
+	<members>
+		<member name="color" type="String" setter="set_color" getter="get_color" default="&quot;COLOR_0&quot;">
+			Targets [constant Mesh.ARRAY_COLOR]
+			Supported attributes: vec3, vec4
```
Comment (Calinou) [inline modules/gltf/doc_classes/GLTFAttributeMap.xml]: Use single-line formatting to make the list more compact:

```suggestion
			Targets [constant Mesh.ARRAY_COLOR]. Supported attributes: vec3, vec4.
```

(Also apply this to other members in the class.)

File: modules/gltf/doc_classes/GLTFDocumentExtension.xml [hunk 14]
```diff
@@ -112,6 +112,15 @@
 				Returns an array of the glTF extensions supported by this GLTFDocumentExtension class. This is used to validate if a glTF file with required extensions can be loaded.
 			</description>
 		</method>
+		<method name="_import_get_attribute_map" qualifiers="virtual">
+			<return type="GLTFAttributeMap" />
+			<param index="0" name="state" type="GLTFState" />
+			<param index="1" name="mesh_index" type="int" />
+			<description>
+				Part of the import process. Allows any mesh to be processed with a custom [GLTFAttributeMap].
+				Return null to ignore.
```
Comment (Calinou) [inline modules/gltf/doc_classes/GLTFDocumentExtension.xml]:
```suggestion
				Part of the import process. Allows any mesh to be processed with a custom [GLTFAttributeMap]. Make this function return [code]null[/code] to ignore.
```

Comment (aaronfranke) [reply modules/gltf/doc_classes/GLTFDocumentExtension.xml]: All GLTFDocumentExtension functions are optional to override, this text seems to hint that overriding it and returning null is required to ignore, when you can also not override the function at all to ignore.

**Comment by lyuma** [comment]
Some thoughts from the asset meeting.

I can see that this feature could be useful for advanced users. We need to balance this with the additional complexity, and the lack of discoverability of this feature.

Since #96748 was approved, this change will have to be rebased on top of it. In particular, it may impact the order of the UV attributes based on `_gltf_primary_texture_coord` and `_gltf_secondary_texture_coord` meta in each of the Material3D attached to the various mesh primitives, often leading TEXCOORD_0 and TEXCOORD_1 to be swapped.

So we can automatically do this fix on the default GLTFAttributeMap. We need to consider how this change may impact extensions that set a custom one. 

Finally, my other concern is this feature is difficult to discover and therefore has a very limited set of users who will know about or use this. Ideally, it would be visible in the advanced importer for example something you can set on each mesh/material to override the GLTFAttributeMap resource slot, and then of course can be again overridden by extensions.

**Comment by Christakou** [comment]
Just wondering if there's been progress on this as I just encountered a usecase in my engine usage where this would be super useful
